server/job-workers/shared-memory-manager.cpp

```diff
@@ -0,0 +1,189 @@
+// Compiler for PHP (aka KPHP)
+// Copyright (c) 2021 LLC «V Kontakte»
+// Distributed under the GPL v3 License, see LICENSE.notice.txt
+
+#include <sys/mman.h>
+
+#include "common/allocators/freelist.h"
+
+#include "server/job-workers/job-stats.h"
+#include "server/job-workers/job-workers-context.h"
+#include "server/php-engine-vars.h"
+
+#include "runtime/critical_section.h"
+
+#include "server/job-workers/shared-memory-manager.h"
+
+namespace job_workers {
+
+struct WorkerProcessMeta {
+  // We can use only 4 messages at once for one worker process:
+  //    mutable request data
+  //    immutable request data
+  //    ok response
+  //    error response (only for job workers)
+  std::array<JobSharedMessage *, 4> attached_messages{};
+
+  void attach(JobSharedMessage *message) noexcept {
+    replace(nullptr, message);
+  }
+
+  void detach(JobSharedMessage *message) noexcept {
+    replace(message, nullptr);
+  }
+
+private:
+  void replace(JobSharedMessage *old_message, JobSharedMessage *new_message) noexcept {
+    for (auto &message : attached_messages) {
+      if (message == old_message) {
+        message = new_message;
+        return;
+      }
+    }
+    assert(false);
+  }
+};
+
+struct alignas(8) SharedMemoryManager::ControlBlock {
+  ControlBlock() noexcept {
+    freelist_init(&free_messages);
+    for (auto &free_mem : free_extra_memory) {
+      freelist_init(&free_mem);
+    }
+  }
+
+  JobStats stats;
+  std::array<WorkerProcessMeta, MAX_WORKERS> workers_table{};
+  freelist_t free_messages{};
+
+  //  index => (1 << index) MB:
+  //    0 => 1MB, 1 => 2MB, 2 => 4MB, 3 => 8MB, 4 => 16MB, 5 => 32MB, 6 => 64MB
+  std::array<freelist_t, JOB_EXTRA_MEMORY_BUFFER_BUCKETS> free_extra_memory{};
+};
+
+void SharedMemoryManager::init() noexcept {
+  static_assert(sizeof(ControlBlock) % 8 == 0, "check yourself");
+  assert(!control_block_);
+
+  const size_t processes = vk::singleton<job_workers::JobWorkersContext>::get().job_workers_num + std::max(workers_n, 1);
```

Comment:
why do we need std::max here?

Reply:
Because the default value of `workers_n` is 0, so it just a trick, if the number of workers is not set. (I have no idea how it would work for 0 workers and N job workers..)